lib/spack/spack/binary_distribution.py

```diff
@@ -2413,3 +2369,118 @@ def __call__(self, spec, **kwargs):
             # Matching a spec constraint
             matches = [s for s in self.possible_specs if s.satisfies(spec)]
         return matches
+
+
+class FetchIndexError(Exception):
+    def __str__(self):
+        if len(self.args) == 1:
+            return str(self.args[0])
+        else:
+            return "{}, due to: {}".format(self.args[0], self.args[1])
+
+
+FetchIndexResult = collections.namedtuple("FetchIndexResult", "etag hash data fresh")
+
+
+class DefaultIndexFetcher:
+    def __init__(self, url, local_hash, urlopen=web_util.urlopen):
+        self.url = url
+        self.local_hash = local_hash
+        self.urlopen = urlopen
+
+    def conditional_fetch(self):
+        # Do an intermediate fetch for the hash
+        # and a conditional fetch for the contents
+        url_index_hash = url_util.join(self.url, _build_cache_relative_path, "index.json.hash")
+
+        try:
+            response = self.urlopen(urllib.request.Request(url_index_hash))
+        except urllib.error.URLError as e:
+            raise FetchIndexError("Could not fetch {}".format(url_index_hash), e) from e
+
+        # Validate the hash
+        remote_hash = response.read(64)
+        if not re.match(rb"[a-f\d]{64}$", remote_hash):
+            raise FetchIndexError("Invalid hash format in {}".format(url_index_hash))
+        remote_hash = remote_hash.decode("utf-8")
+
+        # No need to update further
+        if self.local_hash and remote_hash == self.local_hash:
+            return FetchIndexResult(etag=None, hash=None, data=None, fresh=True)
+
+        # Otherwise, download index.json
+        url_index = url_util.join(self.url, _build_cache_relative_path, "index.json")
+
+        try:
+            response = self.urlopen(urllib.request.Request(url_index))
+        except urllib.error.URLError as e:
+            raise FetchIndexError("Could not fetch index from {}".format(url_index), e)
+
+        try:
+            result = codecs.getreader("utf-8")(response).read()
+        except ValueError as e:
+            return FetchCacheError("Remote index {} is invalid".format(url_index), e)
+
+        computed_hash = compute_hash(result)
+
+        if computed_hash != remote_hash:
+            raise FetchIndexError(
+                "Remote hash {} of {} does not match computed hash {}".format(
+                    remote_hash, url_index, computed_hash
+                )
+            )
```

Comment:
This can be caused by race conditions, we should not error on the reading end if the computed vs. remote hashes are different. The `computed_hash` should be persisted in the cache.

Reply:
Yeah, I agree that keeping `index.json.hash` as a cache-identifier only is the way. But on develop it is also used to verify the integrity of index.json.

Reply:
Ok, I think I'll change that behavior to make `index.json.hash` a cache-identifier only.